cuda/matrix/dense_kernels.cu

```diff
@@ -346,20 +349,214 @@ void compute_norm2(std::shared_ptr<const CudaExecutor> exec,
+    auto num_cols = result->get_size()[1];
+
+    auto row_idxs = result->get_row_idxs();
+    auto col_idxs = result->get_col_idxs();
+    auto values = result->get_values();
+
+    auto stride = source->get_stride();
+
+    auto nnz_prefix_sum = Array<size_type>(exec, num_rows);
+    calculate_nonzeros_per_row(exec, source, &nnz_prefix_sum);
+
+    const size_type grid_dim = ceildiv(num_rows, default_block_size);
+    auto add_values = Array<size_type>(exec, grid_dim);
+
+    kernel::start_prefix_sum<default_block_size, size_type>
+        <<<grid_dim, default_block_size>>>(
+            num_rows, as_cuda_type(nnz_prefix_sum.get_data()),
+            as_cuda_type(add_values.get_data()));
+
+    kernel::finalize_prefix_sum<size_type><<<grid_dim, default_block_size>>>(
+        num_rows, as_cuda_type(nnz_prefix_sum.get_data()),
+        as_cuda_type(add_values.get_data()));
+
+    kernel::fill_in_coo<<<grid_dim, default_block_size>>>(
+        num_rows, num_cols, stride,
+        as_cuda_type(nnz_prefix_sum.get_const_data()),
+        as_cuda_type(source->get_const_values()), as_cuda_type(row_idxs),
+        as_cuda_type(col_idxs), as_cuda_type(values));
+
+    nnz_prefix_sum.clear();
+    add_values.clear();
+}
 
 GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_DENSE_CONVERT_TO_COO_KERNEL);
 
 
+namespace kernel {
+
+
+template <typename ValueType, typename IndexType>
+__global__ __launch_bounds__(default_block_size) void fill_in_csr(
+    size_type num_rows, size_type num_cols, size_type stride,
+    const ValueType *__restrict__ source,
+    const size_type *__restrict__ nnz_prefix_sum,
+    IndexType *__restrict__ row_ptrs, IndexType *__restrict__ col_idxs,
+    ValueType *__restrict__ values)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    if (tidx <= num_rows) {
+        row_ptrs[tidx] = nnz_prefix_sum[tidx];
+    }
+
+    if (tidx < num_rows) {
+        auto write_to = row_ptrs[tidx];
+        for (auto i = 0; i < num_cols; i++) {
+            if (source[stride * tidx + i] != zero<ValueType>()) {
+                values[write_to] = source[stride * tidx + i];
+                col_idxs[write_to] = i;
+                write_to++;
+            }
+        }
+    }
+}
+
+
+}  // namespace kernel
+
+
 template <typename ValueType, typename IndexType>
 void convert_to_csr(std::shared_ptr<const CudaExecutor> exec,
                     matrix::Csr<ValueType, IndexType> *result,
-                    const matrix::Dense<ValueType> *source) NOT_IMPLEMENTED;
+                    const matrix::Dense<ValueType> *source)
+{
+    auto num_rows = result->get_size()[0];
+    auto num_cols = result->get_size()[1];
+
+    auto row_ptrs = result->get_row_ptrs();
+    auto col_idxs = result->get_col_idxs();
+    auto values = result->get_values();
+
+    auto stride = source->get_stride();
+
+    auto nnz_prefix_sum = Array<size_type>(exec, num_rows + 1);
+    calculate_nonzeros_per_row(exec, source, &nnz_prefix_sum);
```

Comment:
I just looked at the `Array` implementation, and I didn't find a easy way of masking `row_idxs` as an `Array` that will not be deleted. This solution should be fine as it is right now.

Reply:
I don't know the exact context here (didn't look at the code in detail), but aren't you describing exactly what [`Array::view`](https://github.com/ginkgo-project/ginkgo/blob/develop/include/ginkgo/core/base/array.hpp#L254) does?

Reply:
True, you are correct, it would work here (I don't know why I overlooked it, I thought about using an `Array` while setting the deleter to do nothing, but I thought it would be too much of a hack).

This will create an array which points to `row_idxs`. You should be able to use it in `calculate_nonzeros_per_row` without any issue (you don't / shouldn't need to clear it afterwards because it would have no effect).
```c++
auto row_idxs_array = Array<size_type>(exec, num_rows + 1, row_idxs);
```

Reply:
Isn't it safer if you call it with `Array::view(...)`? If you use this constructor then you will delete the data afterwards (the `default_deleter` is the `executor_deleter`), is this what you want?

I also did not go through the details of this bits of code, so I may be wrong.

Reply:
Yes...
I forgot to put in the function `view` in my response.
This should work:
```c++
auto row_idxs_array = Array<size_type>::view(exec, num_rows + 1, row_idxs);
calculate_nonzeros_per_row(exec, source, &row_idxs_array);
```
Sorry

Reply:
> I thought about using an `Array` while setting the deleter to do nothing, but I thought it would be too much of a hack

This was the exact use-case why we added deleters to arrays :smile:, and is exactly how views are implemented.

Btw, the use-case comes from the need to efficiently interface with other software, there's an example of why views are useful here: https://github.com/ginkgo-project/ginkgo/blob/develop/examples/3pt_stencil/3pt_stencil.cpp

Reply:
@gflegar 
I know it is implemented that way, but I thought the use case was, for example, add a shared pointer version with array or similar (in general, have more control over the memory management).